src/nvim/message.c

```diff
@@ -3145,6 +3172,65 @@ void msg_check(void)
   }
 }
 
+bool msg_enable_ext(void)
+{
+  return ui_has(kUIMessages) || msg_enable_msgfunc();
+}
+
+bool msg_enable_msgfunc(void)
+{
+  // msgfunc is disabled in command line mode
```

Comment:
For "msgfunc is disabled in command line mode" to be true, you have to add `!(State & CMDLINE)` to the condition below, which makes sense I assume.

Adding that condition also will avoid sending data to msgfunc while the user is typing something in the command line.

Reply:
Why I cannot use `!(State & CMDLINE)`, it is too late for command line output.
The mode change is `in` command line mode.  It is not `before` command line mode.

Reply:
Please see below code.

```c
  redir_off = true;             // don't redirect the typed command
  if (!cmd_silent) {
    gotocmdline(true);
    redrawcmdprompt();          // draw prompt or indent
    ccline.cmdspos = cmd_startcol();
    if (!msg_scroll) {
      msg_ext_clear(false);
    }
  }
  s->xpc.xp_context = EXPAND_NOTHING;
  s->xpc.xp_backslash = XP_BS_NONE;
#ifndef BACKSLASH_IN_FILENAME
  s->xpc.xp_shell = false;
#endif

  if (ccline.input_fn) {
    s->xpc.xp_context = ccline.xp_context;
    s->xpc.xp_pattern = ccline.cmdbuff;
    s->xpc.xp_arg = ccline.xp_arg;
  }

  // Avoid scrolling when called by a recursive do_cmdline(), e.g. when
  // doing ":@0" when register 0 doesn't contain a CR.
  msg_scroll = false;

  State = CMDLINE;

```

`State` is changed in above line, but `redrawcmdprompt()` is called before the mode change.
Should I change `State` set timing??

Reply:
Fixed.